xmodule/capa_block.py

```diff
@@ -2222,6 +2332,102 @@ def score_from_lcp(self, lcp):
         return Score(raw_earned=lcp_score['score'], raw_possible=lcp_score['total'])
 
 
+class GradingMethodHandler:
+    """
+    A class for handling grading method and calculating scores.
+
+    This class allows for flexible handling of grading methods, including options
+    such as considering the last score, the first score, the highest score,
+    or the average score.
+
+    Attributes:
+        - score (Score): The current score.
+        - grading_method (str): The chosen grading method.
+        - score_history (list[Score]): A list to store the history of scores.
+        - max_score (int): The maximum possible score.
+        - mapping_method (dict): A dictionary mapping the grading
+            method to the corresponding handler.
+
+    Methods:
+        - get_score(): Retrieves the updated score based on the grading method.
+        - handle_last_score(): Handles the last score method.
+        - handle_first_score(): Handles the first score method.
+        - handle_highest_score(): Handles the highest score method.
+        - handle_average_score(): Handles the average score method.
+    """
+
+    def __init__(
+        self,
+        score: Score,
+        grading_method: str,
+        score_history: list[Score],
+        max_score: int,
+    ):
+        self.score = score
+        self.grading_method = grading_method
+        self.score_history = score_history
+        self.max_score = max_score
+        self.mapping_method = {
+            GRADING_METHOD.LAST_SCORE: self.handle_last_score,
+            GRADING_METHOD.FIRST_SCORE: self.handle_first_score,
+            GRADING_METHOD.HIGHEST_SCORE: self.handle_highest_score,
+            GRADING_METHOD.AVERAGE_SCORE: self.handle_average_score,
+        }
+
+    def get_score(self) -> Score:
+        """
+        Retrieves the updated score based on the grading method.
+
+        Returns:
+            - Score: The updated score based on the chosen grading method.
+        """
+        return self.mapping_method[self.grading_method]()
+
+    def handle_last_score(self) -> Score:
+        """
+        Retrieves the score based on the last score.
+        It is the last score in the score history.
+
+        Returns:
+            - Score: The score based on the last score.
+        """
+        return self.score_history[-1] if self.score_history else self.score
```

Comment:
In what cases is self.score_history empty or none?

Reply:
When the attempts are reset to zero, and then a rescoring is performed, the `self.score_history` field will be empty.

However, due to this comment I had a doubt whether we should reset the new history fields (`score_history`, `correct_map_history` and `student_answers_history`) at the time of resetting the attempts to zero. Considering that the score is kept, I would think that we should also keep the history. Also, if we keep the history, the if-else would no longer be necessary.

Reply:
Thanks, @BryanttV. I went ahead and addressed your question [here](https://github.com/openedx/edx-platform/pull/33911#discussion_r1507908236).

Reply:
@mariajgrimaldi, to avoid if-else I added the following code:
```python
def get_score(self) -> Score:
    """
    Retrieves the updated score based on the grading method.
    
    Returns:
        - Score: The updated score based on the chosen grading method.
    """
    if len(self.score_history) < 2:
        return self.score
    return self.mapping_method[self.grading_method]()
```
In this way we can eliminate the conditionals within each handler

Reply:
Could you add a comment explaining why we're using < 2? It might not be that obvious

Reply:
In the end I did not add that code because it generated an error in the rescoring, so I followed your suggestion and now I validate it in the constructor.